### PR TITLE
Branchless sign extend in std.bitmanip

### PR DESCRIPTION
**Pull request by deadalnix**
I noticed the codegen was really bad for signed integer, and no doubt, it is using a branch instead of leveraging the capability of the CPU to the an arithmetic right shift. The double shift is trivially turned into an and for unsigned operands, so it doesn't change anything for these with any half decent optimizer involved.

### CODE REVIEW

**Comment by deadalnix** [comment]
So, runnign digger as suggested by the bot plain fail.

```
Building package digger in /home/amaury/.dub/packages/digger-3.0.8/digger/
Performing "debug" build using /home/amaury/dlang/dmd-2.100.1-sym3/linux/bin64/dmd for x86_64.
digger 3.0.8: building configuration "application"...
../../.dub/packages/digger-3.0.8/digger/repo.d(56,18): Deprecation: `repo.DiggerManager.getCallbackCommand` is overriding the deprecated method `ae.sys.d.manager.DManager.getCallbackCommand`
../../.dub/packages/digger-3.0.8/digger/bisect.d(201,14): Deprecation: alias `ae.sys.git.Repository` is deprecated
../../.dub/packages/digger-3.0.8/digger/bisect.d(211,10): Deprecation: alias `ae.sys.git.Repository` is deprecated
std/typecons.d(8622,22): Error: unrecognized trait `classInstanceAlignment`
../../.dub/packages/ae-0.0.3227/ae/sys/datamm.d(34,9): Error: template instance `std.typecons.scoped!(MmFile)` error instantiating
std/typecons.d(8622,22): Error: unrecognized trait `classInstanceAlignment`
../../.dub/packages/ae-0.0.3227/ae/net/nntp/listener.d(31,9): Error: template instance `std.typecons.scoped!(NntpClient)` error instantiating
/home/amaury/dlang/dmd-2.100.1-sym3/linux/bin64/dmd failed with exit code 1.
```

https://github.com/dlang/phobos/pull/8663/checks?check_run_id=10565893255 and alike do swallow the test output, because OF COURSE THEY DO, and so this is impossible to debug.

https://app.circleci.com/pipelines/github/dlang/phobos/6443/workflows/91bb4b81-a2b9-45f0-bdfb-2b86cb414df5/jobs/15970 doesn't swallow the error. Wait a minute, it uses DMD 2.095 and generate a stupid error that doesn't show up in other CI run, and doesn't show up in godbolt when copying and pasting the code: https://godbolt.org/z/bca9G53Gs 

WHAT THE FUCK IS THIS SHIT? DOES ANYTHING WORK AT ALL?

How does one even build phobos now that someone had the brilliant idea to break `make -f posix.mak` ?

**Comment by deadalnix** [comment]
It's a fractal of broken shit to the point you can't even fix one simple thing anymore.

**Comment by ibuclaw** [comment]
> Circle CI's compiler is definitively busted.
> 
> ```d
>     ushort ee = 1028;
>     ee <<= 5U;
>     ee >>= 5U;
>     writeln(ee);
> ```
> 
> Prints 64516 instead of 1028 see ( https://godbolt.org/z/TcbjP76fW ).

2.079 IIRC.

**Comment by maxhaton** [comment]
> > Circle CI's compiler is definitively busted.
> > ```d
> >     ushort ee = 1028;
> >     ee <<= 5U;
> >     ee >>= 5U;
> >     writeln(ee);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Prints 64516 instead of 1028 see ( https://godbolt.org/z/TcbjP76fW ).
> 
> 2.079 is the bootstrap IIRC, but that shouldn't have an effect on runtime codegen (possibly CTFE though)

This has been broken for as long as I can test

**Comment by maxhaton** [comment]
> New dmd backend issue then, gdc prints 1028.

I filed one, I see an issue (referenced in my bug report) from 2009 apparently causing a regression with shifts.

In this case  it seems like dmd movzxs the wrong register entirely

**Comment by deadalnix** [comment]
What's the plan to get the compiler updated on CircleCI? If there is no plan, then I suggest we jettison that pipeline. If the CI is broken, it's not helpful.

**Comment by ibuclaw** [comment]
> What's the plan to get the compiler updated on CircleCI? If there is no plan, then I suggest we jettison that pipeline. If the CI is broken, it's not helpful.

It's not a circleci problem.

https://issues.dlang.org/show_bug.cgi?id=23618

**Comment by deadalnix** [comment]
@ibuclaw it is a CI problem. The bug you point out has already been fixed, but the CI pipeline uses a busted version of the compiler.

**Comment by Herringway** [comment]
> @ibuclaw it is a CI problem. The bug you point out has already been fixed, but the CI pipeline uses a busted version of the compiler.

It might help your case to link to the closed bug report. It does not seem to be fixed in any recent dmd release.

**Comment by ibuclaw** [comment]
> @ibuclaw it is a CI problem. The bug you point out has already been fixed, but the CI pipeline uses a busted version of the compiler.

No it doesn't, and the bug report is not fixed.

**Comment by dkorpel** [comment]
@deadalnix The bit shift bug is now fixed in both DMD master and stable. However, Iain tested this on gdc and ldc and they give assert errors on the same line (bitmanip.d:570), so it seems like the PR has a bug.

File: std/bitmanip.d [hunk 1]
```diff
@@ -121,15 +119,11 @@ private template createAccessors(
...
-                ~ myToString(offset) ~ ";"
-                ~ (T.min < 0
-                   ? "if (result >= " ~ myToString(signBitCheck)
-                   ~ ") result |= " ~ myToString(extendSign) ~ ";"
-                   : "")
-                ~ " return cast("~T.stringof~") result;}\n"
+            enum createAccessors = "@property "~T.stringof~" "~name~"() @safe pure nothrow @nogc const { auto result = cast("~T.stringof~") ("
+                ~ store ~ " >>" ~ myToString(offset) ~ ");"
+                ~ "result <<= " ~ myToString(SignShift) ~ ";"
+                ~ "result >>= " ~ myToString(SignShift) ~ ";"
```
Comment (ibuclaw) [inline std/bitmanip.d:125]: This line is your bug, wrong kind of shift.

Comment (ibuclaw) [inline std/bitmanip.d:125]: This line is causing pipelines to fail. You're using the wrong kind of right shift.

**Review by ibuclaw — CHANGES_REQUESTED** [review]

File: std/bitmanip.d [hunk 2]
```diff
@@ -121,15 +119,11 @@ private template createAccessors(
         else
         {
             // getter
-            enum createAccessors = "@property "~T.stringof~" "~name~"() @safe pure nothrow @nogc const { auto result = "
-                ~"("~store~" & "
-                ~ myToString(maskAllElse) ~ ") >>"
-                ~ myToString(offset) ~ ";"
-                ~ (T.min < 0
```
Comment (ibuclaw) [inline std/bitmanip.d:128]: Removing this line is almost certainly a cause of red pipelines in this PR too.  For instance `enum E { a, b, c }` is typed "int", and `E : 2` does not fit all values unless you treat it as being logically unsigned.

Comment (ibuclaw) [reply std/bitmanip.d:128]: So at the very least you need to have:
```
enum rshiftOp = (T.min < 0 ? ">>= " : ">>>= ");
enum createAccessors = "...."
    ~ "result " ~ rshiftOp ~ myToString(SignShift) ~ ";"
```

Then I imagine all unittests will pass.

Comment (deadalnix) [reply std/bitmanip.d:128]: Interesting. I missed the code special cases enums.

**Comment by deadalnix** [comment]
Check for `T.min < 0` and only sign extend if negative value are possible, even if the underlying type is signed.

**Comment by deadalnix** [comment]
And obviously, github doesn't reset the change requested status, because of course it doesn't. Why would change requested status change once the requested changes have been made?

**Comment by deadalnix** [comment]
And I simplified the code somewhat and fixed the style. This should be good to go.

**Comment by ibuclaw** [comment]
> And obviously, github doesn't reset the change requested status, because of course it doesn't. Why would change requested status change once the requested changes have been made?

It's manual, expand the changes requested dropdown, then the three dots.

image

**Review by ibuclaw — APPROVED** [review]

**Review by RazvanN7 — APPROVED** [review]

**Comment by Herringway** [comment]
This appears to have introduced a regression. https://issues.dlang.org/show_bug.cgi?id=24095

**Pull request merged**